homeworks/hw12/src/bank.rs

```diff
@@ -0,0 +1,999 @@
+
+const MONEY_ZERO: Money = 0.0;
+
+type TransactionId = String;
+
+#[derive(Default)]
+pub struct Bank {
+    accounts: HashMap<String, RefCell<Money>>,
+    accounts_history: HashMap<String, Vec<TransactionId>>,
+    history: BTreeMap<TransactionId, Operation>,
+    ulid_generator: ulid::Generator,
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
+pub struct Operation {
+    #[allow(unused)]
+    id: String,
+    source_account: String,
+    amount: Money,
+    operation_type: OperationType,
+}
+
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+pub enum OperationType {
+    CreateAccount,
+    Deposit,
+    Withdraw,
+    Transfer { target_account: String },
+}
+
+#[derive(Debug, Error, PartialEq)]
+#[error("Account already exists")]
+pub struct AccountDuplicationError {
+    account: String,
+}
+
+#[derive(Debug, Error, PartialEq)]
+#[error("Account does not exist")]
+pub struct AccountNotFoundError {
+    account: String,
+}
+
+#[macro_export]
+macro_rules! account_not_found_error {
+    ($account: expr) => {{
+        error!("Account {} does not exist", $account);
+        Err(AccountNotFoundError { account: $account }.into())
+    }};
+}
+
+#[macro_export]
+macro_rules! check_account_exists {
+    ($self: expr , $account: expr) => {{
+        if !$self.accounts.contains_key($account.as_str()) {
+            error!("Account {} does not exist", $account);
+            return Err(AccountNotFoundError { account: $account }.into());
+        }
+    }};
+}
+
+#[derive(Debug, Error, PartialEq)]
+#[error("Cannot transfer to the same account")]
+pub struct SomeAccountTransferError {
+    account: String,
+}
+
+#[derive(Debug, Error, PartialEq)]
+#[error("Amount must be positive")]
+pub struct AmountNegativeError {
+    account: String,
+    amount: Money,
+}
+
+#[derive(Debug, Error, PartialEq)]
+#[error("Insufficient funds")]
+pub struct InsufficientFundsError {
+    account: String,
+    amount: Money,
+    balance: Money,
+}
+
+#[derive(Debug, Error, PartialEq)]
+pub enum BankError {
+    #[error("Account already exists")]
+    AccountDuplication(#[from] AccountDuplicationError),
+    #[error("Amount must be positive")]
+    AmountNegative(#[from] AmountNegativeError),
+    #[error("Account does not exist")]
+    AccountNotFound(#[from] AccountNotFoundError),
+    #[error("Insufficient funds")]
+    InsufficientFunds(#[from] InsufficientFundsError),
+    #[error("Cannot transfer to the same account")]
+    SomeAccountTransfer(#[from] SomeAccountTransferError),
+}
+
+impl Bank {
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    fn get_next_id(&mut self) -> String {
+        self.ulid_generator
+            .generate_with_source(&mut StdRng::from_entropy())
+            .unwrap()
+            .to_string()
+    }
+
+    fn push_transaction(&mut self, operation: Operation) -> Result<(), BankError> {
+        if !self
+            .accounts
+            .contains_key(operation.source_account.as_str())
```

Comment:
```suggestion
            .contains_key(&operation.source_account)
```

На практике обычно так встречается. Давно уже `as_str()` не видел)